Ported the gdb backtrace from NAV2 (#4190)
ros2/ros2_documentation: source/Tutorials/Intermediate/Get-Backtrace-in-ROS2.rst

> .. redirect-from::
> 
>     Tutorials/Backtrace
> 

Reviewer: This is a new tutorial, so we don't need redirects.  Just remove this.

```suggestion
```

> Getting Backtraces in ROS 2
> =================================

Reviewer: ```suggestion
Getting Backtraces in ROS 2
===========================
```

.. contents:: Table of Contents
   :depth: 2
   :local:

> **Goal:**  Methods for getting Backtrace in ROS 2 

Reviewer 2: ```suggestion
**Goal:**  Methods for getting backtrace in ROS 2 
```

**Tutorial level:** Intermediate

**Time:** 15 minutes

> The following steps show ROS 2 users how to get traces from specific servers when they encounter a problem..

Reviewer: I'm not sure what a `specific server` is here (and elsewhere).  Are these processes?

Author: I guess they are referring to things like planner server, controller Server etc. in NAV2
Shall I remove this point ?
@clalancette @fujitatomoya

Overview
========

This document explains one set of methods for getting backtraces for ROS 2.
There are many ways to accomplish this, but this is a good starting point for new C++ developers without GDB experience.

> The following steps show ROS 2 users how to get traces from specific servers when they encounter a problem.

Reviewer 2: same here for `specific servers`, just pointing not to miss.

This tutorial applies to both simulated and physical robots.

This will cover how to get a backtrace from a specific node using ``ros2 run``, from a launch file representing a single node using ``ros2 launch``, and from a more complex orchestration of nodes.
By the end of this tutorial, you should be able to get a backtrace when you notice a server crashing in ROS 2.

Preliminaries
=============

> GDB is the most popular debugger for C++ on Unix systems.

Reviewer 2: ```suggestion
GDB is the most popular debugger for C/C++ on Unix systems.
```

It can be used to determine the reason for a crash and track threads.
It may also be used to add breakpoints in your code to check values in memory a particular points in your software.

Using GDB is a critical skill for all software developers working on C/C++.
Many IDEs will have some kind of debugger or profiler built in, but with ROS, there are few IDEs to choose.
Therefore it's important to understand how to use these raw tools you have available rather than relying on an IDE to provide them.
Further, understanding these tools is a fundamental skill of C/C++ development and leaving it up to your IDE can be problematic if you change roles and no longer have access to it or are doing development on the fly through an ssh session to a remote asset.

Using GDB luckily is fairly simple after you have the basics under your belt.
The first step is to add ``-g`` to your compiler flags for the ROS package you want to profile / debug.

> This flag builds debug symbols that GDB and valgrind can read to tell you specific lines of code in your project are failing and why.

Reviewer: I think we should remove the reference to `valgrind` here, as we don't discuss it otherwise.

If you do not set this flag, you can still get backtraces but it will not provide line numbers for failures.

> Be sure to remove this flag after debugging, it will slow down performance at run-time.

Reviewer: This is actually not true; using `-g` will just generate the debugging information, which isn't used at runtime unless you are using a debugger.  So I'd remove this line.

Adding the following line to your ``CMakeLists.txt`` for your project should do the trick.
If your project already has a ``add_compile_options()``, you can simply add ``-g`` to it.
Then simply rebuild your workspace with this package ``colcon build --packages-select <package-name>``.
It may take a little longer than usual to compile.

.. code-block:: cmake

>   add_compile_options(-g)

Reviewer 2: instead of modifying CMakeLists.txt, what about introducing `--cmake-args -DCMAKE_BUILD_TYPE=Debug` to rebuild the packages to debug? it would be easier for most cases?

Author: yes, I feel that would be more convenient and streamlined way while debugging  in many cases. though we do have the option of modifying CMakeLists.txt for certain scenarios where you need the debug flag `(-g)` specifically and want to be very precise about other compiler options, this might be a better choice. 

Shall we keep both in the documentation, having `--cmake-args -DCMAKE_BUILD_TYPE=Debug` as default and adding `CMakeLIst.txt` as an option in the Notes section.

Reviewer 3: In the original issue, I proposed using mixins.

Now you're ready to debug your code!
If this was a non-ROS project, at this point you might do something like below.
Here we're launching a GDB session and telling our program to immediately run.
Once your program crashes, it will return a gdb session prompt denoted by ``(gdb)``.
At this prompt you can access the information you're interested in.
However, since this is a ROS project with lots of node configurations and other things going on, this isn't a great option for beginners or those that don't like tons of commandline work and understanding the filesystem.

.. code-block:: bash

  gdb ex run --args /path/to/exe/program

Below are sections to describe the 3 major situations you could run into with ROS 2-based systems.
Read the section that best describes the problem you're attempting to solve.

> From a Node

Reviewer 2: This sounds weird for me, it is supposed to be executable but node. besides, in the executable there could be multiple nodes running in the same process space.

How about `ros2 run --prefix`?

===========

Just as in our non-ROS example, we need to setup a GDB session before launching our ROS 2 node.

> While we could set this up through the commandline with some knowledge of the ROS 2 file system, we can instead use the launch ``--prefix`` option the kind folks at Open Robotics provided for us.

Reviewer 2: I would remove these lines, i think we can just start explaining about `--prefix` option here.

- It crashed in ``at()`` on STL vector line 1091 after throwing an exception from a range check failure.

These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.
Then you can deduce why it crashed.
When you are done with GDB, type ``quit`` and it will exit the session and kill any processes still up.
It may ask you if you want to kill some threads at the end, say yes.

From a Launch File
==================

Just as in our non-ROS example, we need to setup a GDB session before launching our ROS 2 launch file.
While we could set this up through the commandline, we can instead make use of the same mechanics that we did in the ``ros2 run`` node example, now using a launch file.

In your launch file, find the node that you're interested in debugging.
For this section, we assume that your launch file contains only a single node (and potentially other information as well).
The ``Node`` function used in the ``launch_ros`` package will take in a field ``prefix`` taking a list of prefix arguments.
We will insert the GDB snippet here with one change from our node example, use of ``xterm``.

> ``xterm`` will pop up a new terminal window to show and interact with GDB.

Reviewer 2: i think this requires window system. if we are doing this debug remotely, `xterm` could not pop up.

Author: I have mentioned this in the updated documentation by also providing usage and examples for the same.

      parameters=[
        get_package_share_directory("slam_toolbox") + '/config/mapper_params_online_sync.yaml',
        {'use_sim_time': use_sim_time}
      ],
      package='slam_toolbox',
      executable='sync_slam_toolbox_node',
      name='slam_toolbox',
      prefix=['xterm -e gdb -ex run --args'],
      output='screen')

Just as before, this prefix will launch a GDB session, now in ``xterm`` and run the launch file you requested with all the additional launch arguments defined.

Once your server crashes, you'll see a prompt like below, now in the ``xterm`` session. At this point you can now get a backtrace.

.. code-block:: bash

  (gdb)

In this session, type ``backtrace`` and it will provide you with a backtrace.
Copy this for your needs.
See the example trace in the section above for an example.

These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.
Then you can deduce why it crashed.
When you are done with GDB, type ``quit`` and it will exit the session and kill any processes still up.
It may ask you if you want to kill some threads at the end, say yes.

> From Large Project
> ==================

Reviewer: ```suggestion
From a Large Project
====================
```

Working with launch files with multiple nodes is a little different so you can interact with your GDB session without being bogged down by other logging in the same terminal.
For this reason, when working with larger launch files, its good to pull out the specific server you're interested in and launching it seperately.

> These instructions are targeting ROS2, but are applicable to any large project with many nodes of any type in a series of launch file(s).

Reviewer 2: ```suggestion
These instructions are targeting ROS 2, but are applicable to any large project with many nodes of any type in a series of launch file(s).
```

As such, for this case, when you see a crash you'd like to investigate, its beneficial to separate this server from the others.

If your server of interest is being launched from a nested launch file (e.g. an included launch file) you may want to do the following:

- Comment out the launch file inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

- Launch the server's launch file in another terminal following the instructions in `From a Launch File`_.

Alternatively, if you server of interest is being launched in these files directly (e.g. you see a ``Node``, ``LifecycleNode``, or inside a ``ComponentContainer``), you will need to seperate this from the others:

- Comment out the node's inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

- Launch the server's node in another terminal following the instructions in `From a Node`_.

> .. note::
>   Note that in this case, you may need to remap or provide parameter files to this node if it was previously provided by the launch file. Using ``--ros-args`` you can give it the path to the new parameters file, remaps, or names. See `this ROS 2 tutorial <https://docs.ros.org/en/rolling/Guides/Node-arguments.html>`_ for the commandline arguments required.

Reviewer: ```suggestion
.. note::

  Note that in this case, you may need to remap or provide parameter files to this node if it was previously provided by the launch file.
  Using ``--ros-args`` you can give it the path to the new parameters file, remaps, or names.
  See :doc:`this tutorial <../../Guides/Node-arguments.html>` for the commandline arguments required.
```

To debug directly from the nav2 bringup launch files you may want to do the following:

- Add ``prefix=['xterm -e gdb -ex run --args']`` to the non-composed node in the appropriate launch file. 

- Recompile the package of interest with ``-g`` flag for debug symbols.

- Launch normally with ``ros2 launch nav2_bringup tb3_simulation_launch.py use_composition:=False``. A seperate xterm window will open with the proccess of intrest running in gdb.

.. note::
  Turning off composition has serious performance impacts. If this is important to you please follow "From Large Project". 

Once your server crashes, you'll see a prompt like below in the xterm window. At this point you can now get a backtrace.

.. code-block:: bash

  (gdb)

In this session, type ``backtrace`` and it will provide you with a backtrace.
Copy this for your needs.
See the example trace in the section above for an example.

These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.
Then you can deduce why it crashed.
When you are done with GDB, type ``quit`` and it will exit the session and kill any processes still up.

> It may ask you if you want to kill some threads at the end, say yes.

Reviewer: Since this part is Navigation 2 specific, let's remove it completely.

Getting Backtraces in ROS 2
===========================

.. contents:: Table of Contents
   :depth: 2
   :local:

**Goal:**  Methods for getting backtrace in ROS 2 

**Tutorial level:** Intermediate

**Time:** 15 minutes

> The following steps show ROS 2 users how to get traces when they encounter a problem..

Reviewer 3: ```suggestion
The following steps show ROS 2 users how to get traces when they encounter a problem.
```

Can remove double period.

Overview
========

> This document explains one set of methods for getting backtraces for ROS 2.

Reviewer 3: You might want to show what a backtrace is and why it's useful to motivate it more than "when they encounter a problem". Perhaps an embedded screenshot?

There are many ways to accomplish this, but this is a good starting point for new C++ developers without GDB experience.

The following steps show ROS 2 users how to get traces from specific nodes when they encounter a problem.
This tutorial applies to both simulated and physical robots.

This will cover how to get a backtrace from a specific node using ``ros2 run``, from a launch file representing a single node using ``ros2 launch``, and from a more complex orchestration of nodes.
By the end of this tutorial, you should be able to get a backtrace when you notice a node crashing in ROS 2.

Preliminaries
=============

GDB is the most popular debugger for C/C++ on Unix systems.
It can be used to determine the reason for a crash and track threads.

> It may also be used to add breakpoints in your code to check values in memory a particular points in your software.

Reviewer 3: ```suggestion
It may also be used to add breakpoints in your code to check values in memory at particular points in your software.
```

Using GDB is a critical skill for all software developers working on C/C++.
Many IDEs will have some kind of debugger or profiler built in, but with ROS, there are few IDEs to choose.
Therefore it's important to understand how to use these raw tools you have available rather than relying on an IDE to provide them.
Further, understanding these tools is a fundamental skill of C/C++ development and leaving it up to your IDE can be problematic if you change roles and no longer have access to it or are doing development on the fly through an ssh session to a remote asset.

Using GDB luckily is fairly simple after you have the basics under your belt.

> The first step is to add ``-g`` to your compiler flags for the ROS package you want to profile / debug.

Reviewer 3: I would lead with `--cmake-args` as the first option because it doesn't require code changes.

This flag builds debug symbols that GDB can read to tell you specific lines of code in your project are failing and why.
If you do not set this flag, you can still get backtraces but it will not provide line numbers for failures.

Using ``--cmake-args -DCMAKE_BUILD_TYPE=Debug`` in your ``colcon build`` command will add the ``-g`` flag to your build.

.. code-block:: bash

>   colcon build --packages-select <package_name> --cmake-args -DCMAKE_BUILD_TYPE=Debug 

Reviewer 3: ```suggestion
  colcon build --packages-up-to <package_name> --cmake-args -DCMAKE_BUILD_TYPE=Debug 
```

I recommend building the whole stack in debug, in the case that the crash is in a dependency.

.. If your project already has a ``add_compile_options()``, you can simply add ``-g`` to it.
.. Then simply rebuild your workspace with this package ``colcon build --packages-select <package-name>``.
.. It may take a little longer than usual to compile.

Now you're ready to debug your code!
If this was a non-ROS project, at this point you might do something like below.
Here we're launching a GDB session and telling our program to immediately run.
Once your program crashes, it will return a gdb session prompt denoted by ``(gdb)``.
At this prompt you can access the information you're interested in.
However, since this is a ROS project with lots of node configurations and other things going on, this isn't a great option for beginners or those that don't like tons of commandline work and understanding the filesystem.

.. code-block:: bash

  gdb ex run --args /path/to/exe/program

Below are sections to describe the 3 major situations you could run into with ROS 2-based systems. 
Read the section that best describes the problem you're attempting to solve.

Debugging a specific node with GDB
==================================

> To easily set up a GDB session before launching a ROS2 node, leverage the ``--prefix`` option in launch files. This option allows you to specify a command to execute before the node starts. For GDB debugging, use it as follows:

Reviewer 2: one sentence per line.

and this is important, we need a `space` between `ROS` and `2` 😄 https://discourse.ros.org/t/should-i-put-a-space-between-ros-and-2/27949

can you apply the same change to other places in this PR?

```suggestion
To easily set up a GDB session before launching a ROS 2 node, leverage the ``--prefix`` option in launch files.
This option allows you to specify a command to execute before the node starts. For GDB debugging, use it as follows:
```

.. note::

  Important Note: Keep in mind that a ROS2 executable might contain multiple nodes. The ``--prefix`` approach ensures you're debugging the correct node within the process.

**Why Direct GDB Usage Can Be Tricky**

> ``--prefix`` will execute some bits of code before our ros2 command allowing us to insert some information. If you attempted to do ``gdb ex run --args ros2 run <pkg> <node>`` as analog to our example in the preliminaries, you’d find that it couldn’t find the ros2 command. Additionally, trying to source your workspace within GDB would fail for similar reasons. This is because GDB, when launched this way, lacks the environment setup that normally makes the ros2 command available.

Reviewer 3: ```suggestion
``--prefix`` will execute some bits of code before our ros2 command allowing us to insert some information. If you attempted to do ``gdb ex run --args ros2 run <pkg> <node>`` as analog to our example in the preliminaries, you’d find that it couldn’t find the ``ros2`` command. Additionally, trying to source your workspace within GDB would fail for similar reasons. This is because GDB, when launched this way, lacks the environment setup that normally makes the ``ros2`` command available.
```

- It crashed in ``at()`` on STL vector line 1091 after throwing an exception from a range check failure.

These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.
Then you can deduce why it crashed.
When you are done with GDB, type ``quit`` and it will exit the session and kill any processes still up.
It may ask you if you want to kill some threads at the end, say yes.

From a Launch File
==================

Just as in our non-ROS example, we need to setup a GDB session before launching our ROS 2 launch file.
While we could set this up through the commandline, we can instead make use of the same mechanics that we did in the ``ros2 run`` node example, now using a launch file.

In your launch file, find the node that you’re interested in debugging. For this section, we assume that your launch file contains only a single node (and potentially other information as well). 
The ``Node`` function used in the ``launch_ros`` package will take in a field prefix taking a list of prefix arguments. We will insert the GDB snippet here. **Consider the following approaches, depending on your setup:**

> - **Local Debugging with Windowing System:**  If you are debugging locally and have a windowing system available, use:

Reviewer 3: Would `GUI` be more a more appropriate word than `Windowing`

Reviewer: Yeah, I think that would be clearer.

These instructions are targeting ROS 2, but are applicable to any large project with many nodes of any type in a series of launch file(s).

As such, for this case, when you see a crash you'd like to investigate, its beneficial to separate this server from the others.

If your server of interest is being launched from a nested launch file (e.g. an included launch file) you may want to do the following:

- Comment out the launch file inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

- Launch the server's launch file in another terminal following the instructions in `From a Launch File`_.

Alternatively, if you server of interest is being launched in these files directly (e.g. you see a ``Node``, ``LifecycleNode``, or inside a ``ComponentContainer``), you will need to seperate this from the others:

- Comment out the node's inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

> - Launch the server's node in another terminal following the instructions in `From a Node`_.

Reviewer 2: where is this `From a Node` instruction?

.. note::

  Note that in this case, you may need to remap or provide parameter files to this node if it was previously provided by the launch file.
  Using ``--ros-args`` you can give it the path to the new parameters file, remaps, or names.
  See :doc:`this tutorial <../../Guides/Node-arguments.html>` for the commandline arguments required.

  We understand this can be a pain, so it might encourage you to rather have each node possible as a separately included launch file to make debugging easier. An example set of arguments might be ``--ros-args -r __node:=<node_name> --params-file /absolute/path/to/params.yaml`` (as a template).

Once your server crashes, you'll see a prompt like below in the specific server's terminal. At this point you can now get a backtrace.

.. code-block:: bash

  (gdb)

In this session, type ``backtrace`` and it will provide you with a backtrace.
Copy this for your needs.
See the example trace in the section above for an example.

> These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.

Reviewer 3: This stuff is repeated a few times. Can you only put this info once?

Reviewer: Yeah, agreed.  A section at the beginning about "Reading the stack trace" might be a good idea.

Debugging tests with GDB
========================

If a C++ test is failing, GDB can be used directly on the test executable in the build directory.
Ensure to build the code in debug mode.
Since the previous build type may be cached by CMake, clean the cache and rebuild.

.. code-block:: console

  colcon build --cmake-clean-cache --mixin debug

In order for GDB to load debug symbols for any shared libraries called, make sure to source your environment.
This configures the value of ``LD_LIBRARY_PATH``.

.. code-block:: console

  source install/setup.bash

Finally, run the test directly through GDB.
For example:

.. code-block:: console

  gdb -ex run ./build/rcl/test/test_logging

If the code is throwing an unhandled exception, you can catch it in GDB before gtest handles it.

.. code-block:: console

  gdb ./build/rcl/test/test_logging
  catch throw
  run

Automatic backtrace on crash
============================

The `backward-cpp <https://github.com/bombela/backward-cpp>`_ library provides beautiful stack traces, and the `backward_ros <https://github.com/pal-robotics/backward_ros/tree/foxy-devel>`_ wrapper simplifies its integration.

> Just add it as a dependency and `find_package` it in your CMakeLists and the backward libraries will be injected in all your executables and libraries.

Reviewer 2: do we want to introduce this `backward_ros` repository here? it only supports `foxy-devel`? if that can be used with `rolling` `iron` and `humble` branch, i think that is fine.

Reviewer 3: I am ok to leave it out of this first round; GDB alone is very useful.

Reviewer: It looks like `backward_ros` is available on all distributions.  However, I think we should indeed change the link to just https://github.com/pal-robotics/backward_ros .

Getting Backtraces in ROS 2
===========================

.. contents:: Table of Contents
   :depth: 2
   :local:

> **Goal:**  Methods for getting backtrace in ROS 2 

Reviewer: ```suggestion
**Goal:**  Show various methods for getting backtraces in ROS 2 
```

Using GDB luckily is fairly simple after you have the basics under your belt.
The first step is to add ``-g`` to your compiler flags for the ROS package you want to profile / debug.
This flag builds debug symbols that GDB can read to tell you specific lines of code in your project are failing and why.
If you do not set this flag, you can still get backtraces but it will not provide line numbers for failures.

Using ``--cmake-args -DCMAKE_BUILD_TYPE=Debug`` in your ``colcon build`` command will add the ``-g`` flag to your build.

.. code-block:: bash

  colcon build --packages-up-to <package_name> --cmake-args -DCMAKE_BUILD_TYPE=Debug 

this must do the trick.

.. Adding the following line to your ``CMakeLists.txt`` for your project should do the trick.
.. If your project already has a ``add_compile_options()``, you can simply add ``-g`` to it.
.. Then simply rebuild your workspace with this package ``colcon build --packages-select <package-name>``.

> .. It may take a little longer than usual to compile.

Reviewer: ```suggestion
```

Now you're ready to debug your code!
If this was a non-ROS project, at this point you might do something like below.
Here we're launching a GDB session and telling our program to immediately run.
Once your program crashes, it will return a gdb session prompt denoted by ``(gdb)``.
At this prompt you can access the information you're interested in.
However, since this is a ROS project with lots of node configurations and other things going on, this isn't a great option for beginners or those that don't like tons of commandline work and understanding the filesystem.

.. code-block:: bash

  gdb ex run --args /path/to/exe/program

Below are sections to describe the 3 major situations you could run into with ROS 2-based systems. 
Read the section that best describes the problem you're attempting to solve.

Debugging a specific node with GDB
==================================

To easily set up a GDB session before launching a ROS 2 node, leverage the ``--prefix`` option in launch files. 
This option allows you to specify a command to execute before the node starts. 
For GDB debugging, use it as follows:

.. note::

>   Important Note: Keep in mind that a ROS 2 executable might contain multiple nodes. 

Reviewer: ```suggestion
  Keep in mind that a ROS 2 executable might contain multiple nodes. 
```

  The ``--prefix`` approach ensures you're debugging the correct node within the process.

**Why Direct GDB Usage Can Be Tricky**

``--prefix`` will execute some bits of code before our ROS 2 command allowing us to insert some information. 
If you attempted to do ``gdb ex run --args ros2 run <pkg> <node>`` as analog to our example in the preliminaries, you’d find that it couldn’t find the ``ros2`` command. 
Additionally, trying to source your workspace within GDB would fail for similar reasons. 

> This is because GDB, when launched this way, lacks the environment setup that normally makes the ``ROS 2`` command available.

Reviewer: ```suggestion
This is because GDB, when launched this way, lacks the environment setup that normally makes the ``ros2`` command available.
```

**Simplifying the Process with --prefix**

Rather than having to revert to finding the install path of the executable and typing it all out, we can instead use ``--prefix``. 
This allows us to use the same ``ros2 run`` syntax you’re used to without having to worry about some of the GDB details.

.. code-block:: bash

  ros2 run --prefix 'gdb -ex run --args' <pkg> <node> --all-other-launch arguments 

**The GDB Experience**

Just as before, this prefix will launch a GDB session and run the node you requested with all the additional command-line arguments. 
You should now have your node running and should be chugging along with some debug printing.

> Once your server crashes, you’ll see a prompt like below. At this point you can now get a backtrace.

Reviewer: ```suggestion
Once your node crashes, you’ll see a prompt like below.
At this point you can get a backtrace.
```

  #10 0x0000555555559cfc in main (argc=1, argv=0x7fffffffc108)
      at /home/steve/Documents/nav2_ws/src/gdb_test_pkg/src/main.cpp:25

In this example you should read this in the following way, starting at the bottom:

- In the main function, on line 25 we call a function VectorCrash.

- In VectorCrash, on line 44, we crashed in the Vector's ``at()`` method with input ``100``.

- It crashed in ``at()`` on STL vector line 1091 after throwing an exception from a range check failure.

These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.
Then you can deduce why it crashed.
When you are done with GDB, type ``quit`` and it will exit the session and kill any processes still up.
It may ask you if you want to kill some threads at the end, say yes.

From a Launch File
==================

Just as in our non-ROS example, we need to setup a GDB session before launching our ROS 2 launch file.
While we could set this up through the commandline, we can instead make use of the same mechanics that we did in the ``ros2 run`` node example, now using a launch file.

> In your launch file, find the node that you’re interested in debugging. For this section, we assume that your launch file contains only a single node (and potentially other information as well). 

Reviewer: ```suggestion
In your launch file, find the node that you’re interested in debugging.
For this section, we assume that your launch file contains only a single node (and potentially other information as well). 
```

The ``Node`` function used in the ``launch_ros`` package will take in a field prefix taking a list of prefix arguments. 
We will insert the GDB snippet here. 

> **Consider the following approaches, depending on your setup:**

Reviewer: ```suggestion

**Consider the following approaches, depending on your setup:**
```

- **Local Debugging with Windowing System:**  If you are debugging locally and have a windowing system available, use:

> .. code-block:: bash

Reviewer: ```suggestion
.. code-block:: python
```

        {'use_sim_time': use_sim_time}
    ],
    package='slam_toolbox',
    executable='sync_slam_toolbox_node',
    name='slam_toolbox',
    prefix=['xterm -e gdb -ex run --args'],  # For interactive GDB in a separate window
    output='screen')

- **Remote Debugging (No Windowing System):** If debugging remotely without a windowing system, omit ``xterm -e`` :

.. code-block:: bash

  prefix=['gdb -ex run --args']

GDB's output and interaction will happen within the terminal session where you launched the ROS 2 application.
Here's an similar example for the ``'start_sync_slam_toolbox_node'`` -

.. code-block:: python

  start_sync_slam_toolbox_node = Node(
    parameters=[
        get_package_share_directory("slam_toolbox") + '/config/mapper_params_online_sync.yaml',
        {'use_sim_time': use_sim_time}
    ],
    package='slam_toolbox',
    executable='sync_slam_toolbox_node',
    name='slam_toolbox',
    prefix=['gdb -ex run --args'],  # For GDB within the launch terminal
    output='screen')

Just as before, this prefix will launch a GDB session, now in ``xterm`` and run the launch file you requested with all the additional launch arguments defined.

> Once your server crashes, you'll see a prompt like below, now in the ``xterm`` session. 

Reviewer: ```suggestion
Once your node crashes, you'll see a prompt like below, now in the ``xterm`` session. 
```

At this point you can now get a backtrace.

.. code-block:: bash

  (gdb)

In this session, type ``backtrace`` and it will provide you with a backtrace.
Copy this for your needs.
See the example trace in the section above for an example.

These traces take some time to get used to reading, but in general, start at the bottom and follow it up the stack until you see the line it crashed on.
Then you can deduce why it crashed.
When you are done with GDB, type ``quit`` and it will exit the session and kill any processes still up.
It may ask you if you want to kill some threads at the end, say yes.

From a Large Project
====================

Working with launch files with multiple nodes is a little different so you can interact with your GDB session without being bogged down by other logging in the same terminal.

> For this reason, when working with larger launch files, its good to pull out the specific server you're interested in and launching it seperately.

Reviewer: ```suggestion
For this reason, when working with larger launch files, its good to pull out the specific node you're interested in and launching it separately.
```

These instructions are targeting ROS 2, but are applicable to any large project with many nodes of any type in a series of launch file(s).

> As such, for this case, when you see a crash you'd like to investigate, its beneficial to separate this server from the others.

Reviewer: Both of these sentences seem unnecessary and redundant with the sentence above.

```suggestion
```

> If your server of interest is being launched from a nested launch file (e.g. an included launch file) you may want to do the following:

Reviewer: ```suggestion
If your node of interest is being launched from a nested launch file (e.g. an included launch file) you may want to do the following:
```

- Comment out the launch file inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

> - Launch the server's launch file in another terminal following the instructions in `From a Launch File`_.

Reviewer: ```suggestion
- Launch the node's launch file in another terminal following the instructions in `From a Launch File`_.
```

Also the `From a Launch File` tutorial doesn't seem to exist.  So we either need to port that over at the same time, or remove this reference.

> Alternatively, if you server of interest is being launched in these files directly (e.g. you see a ``Node``, ``LifecycleNode``, or inside a ``ComponentContainer``), you will need to seperate this from the others:

Reviewer: ```suggestion
Alternatively, if your node of interest is being launched in these files directly (e.g. you see a ``Node``, ``LifecycleNode``, or inside a ``ComponentContainer``), you will need to seperate this from the others:
```

- Comment out the node's inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

- Launch the server's node in another terminal following the instructions in `From a Node`_.

.. note::

>   Note that in this case, you may need to remap or provide parameter files to this node if it was previously provided by the launch file.

Reviewer: ```suggestion
  In this case you may need to remap or provide parameter files to this node if it was previously provided by the launch file.
```

  Using ``--ros-args`` you can give it the path to the new parameters file, remaps, or names.
  See :doc:`this tutorial <../../Guides/Node-arguments.html>` for the commandline arguments required.

  We understand this can be a pain, so it might encourage you to rather have each node possible as a separately included launch file to make debugging easier. 
  An example set of arguments might be ``--ros-args -r __node:=<node_name> --params-file /absolute/path/to/params.yaml`` (as a template).

> Once your server crashes, you'll see a prompt like below in the specific server's terminal. At this point you can now get a backtrace.

Reviewer: ```suggestion
Once your node crashes, you'll see a prompt like below in the specific server's terminal. At this point you can now get a backtrace.
```

> How to get Backtraces in ROS 2

Reviewer 2: @GauravKumar9920 i think what @clalancette  meant is to replace this whole doc under the section https://github.com/ros2/ros2_documentation/tree/rolling/source/How-To-Guides? it is now targeting to https://github.com/ros2/ros2_documentation/tree/rolling/source/Tutorials

@clalancette let me know if i am mistaken.

Getting Backtraces in ROS 2
===========================

.. contents:: Table of Contents
   :depth: 2
   :local:

**Goal:**  Show various methods for getting backtraces in ROS 2 

**Tutorial level:** Intermediate

**Time:** 15 minutes

The following steps show ROS 2 users how to get traces when they encounter a problem.

Overview
========

.. This document explains methods for getting backtraces in ROS 2, which are essential for debugging.  
.. A backtrace provides a snapshot of the entire sequence of function calls that led up to a crash or error.

> .. There are many ways to accomplish this, but this is a good starting point for new C++ developers without GDB experience.

Reviewer: We shouldn't leave these comments around.  Either they should become part of the main narrative, or they should be removed.

**What is a Backtrace ?**

- Imagine your program is like a multi-story building where each floor is a function it's currently executing. 
  A backtrace is like an emergency map that shows you exactly how the program got to the floor where it crashed.
  
- It lists out the sequence of functions that were called, one on top of the other, leading up to the point of failure.

**Why is it Useful?**

- **Pinpoints the Problem:** Instead of guessing where in your code an error occurred, the backtrace shows you the exact line number responsible for the crash.
- **Reveals Context:** You can see the chain of events (functions calling other functions) that ultimately triggered the failure. 
  This helps you understand not just where things went wrong, but also why.

> **Visual Analogy**:  Stack of Pancakes

Reviewer: I'm not a huge fan of having two different analogies in this opening.  Above we talk about it like a "multi-story building", and here we talk about it like a pancake.  We should pick one of those analogies and stick with it.

**How the Backtrace Helps:**

- **Crash Origin:** Shows you the exact line in ``functionC()`` that triggered the crash.
- **Call Sequence:** Reveals that ``main()`` called ``functionA()``, which called ``functionB()``, which ultimately led to the error in ``functionC()``.

The above example gave us a clear picture of what is a backtrace and how it can be useful.
Now, the following steps show ROS 2 users how to get traces from specific nodes when they encounter a problem.
This tutorial applies to both simulated and physical robots.

This will cover how to get a backtrace from a specific node using ``ros2 run``, from a launch file representing a single node using ``ros2 launch``, and from a more complex orchestration of nodes.
By the end of this tutorial, you should be able to get a backtrace when you notice a node crashing in ROS 2.

Preliminaries
=============

GDB is the most popular debugger for C/C++ on Unix systems.
It can be used to determine the reason for a crash and track threads.
It may also be used to add breakpoints in your code to check values in memory at particular points in your software.

Using GDB is a critical skill for all software developers working on C/C++.

> Many IDEs will have some kind of debugger or profiler built in, but with ROS 2, there are few IDEs to choose.
> Therefore it's important to understand how to use these raw tools you have available rather than relying on an IDE to provide them.
> Further, understanding these tools is a fundamental skill of C/C++ development and leaving it up to your IDE can be problematic if you change roles and no longer have access to it or are doing development on the fly through an ssh session to a remote asset.

Reviewer: It isn't actually true that there are "few" IDEs for ROS 2; you can use any IDE with ROS 2.  But it can be tricky to set them up properly.  So I think this should be reworded something like:

```suggestion
While many IDEs have some kind of debugger or profiler built in, it is important to understand how to use these raw tools you have available rather than relying on an IDE to provide them.
Understanding these tools is a fundamental skill of C/C++ development and leaving it up to your IDE can be problematic if you change roles and no longer have access to it or are doing development on the fly through an ssh session to a remote asset.
```

  If you do not set this flag, you can still get backtraces but it will not provide line numbers for failures.

Now you're ready to debug your code!
If this was a non-ROS project, at this point you might do something like below.
Here we're launching a GDB session and telling our program to immediately run.
Once your program crashes, it will return a gdb session prompt denoted by ``(gdb)``.
At this prompt you can access the information you're interested in.
However, since this is a ROS project with lots of node configurations and other things going on, this isn't a great option for beginners or those that don't like tons of commandline work and understanding the filesystem.

.. code-block:: bash

  gdb ex run --args /path/to/exe/program

Below are sections to describe the 3 major situations you could run into with ROS 2-based systems. 
Read the section that best describes the problem you're attempting to solve.

Debugging a specific node with GDB
==================================

To easily set up a GDB session before launching a ROS 2 node, leverage the ``--prefix`` option in launch files. 
This option allows you to specify a command to execute before the node starts. 

> For GDB debugging, use it as follows:

Reviewer: This paragraph seems to be misplaced; it is talking about launch files, but this section is all about using ``ros2 run``.  This should be rewritten.

This allows us to use the same ``ros2 run`` syntax you’re used to without having to worry about some of the GDB details.

.. code-block:: bash

  ros2 run --prefix 'gdb -ex run --args' <pkg> <node> --all-other-launch arguments 

**The GDB Experience**

Just as before, this prefix will launch a GDB session and run the node you requested with all the additional command-line arguments. 
You should now have your node running and should be chugging along with some debug printing.

Reading the Stack Trace
=======================

After you obtain a backtrace using GDB, here's how to interpret it:

- Start at the Bottom: Backtraces list function calls in reverse chronological order. 
  The function at the bottom is where the crash originates.

- Follow the Stack Upwards: Each line above represents the function that called the function below it. 
  Trace upwards until you reach a line of code within your own project. 
  This often reveals where the problem initiated.

- Debugging Clues: Function names and their arguments can provide valuable clues about what went wrong.

> **How to Debug once your Node Crashes-**

Reviewer: ```suggestion
**How to Debug once your Node Crashes**
```

Once your node crashes, you’ll see a prompt like below.
At this point you can get a backtrace.

.. code-block:: bash

  (gdb)

In this session, type ``backtrace`` and it will provide you with a backtrace.
Copy this for your needs.

> **Example backtrace-**

Reviewer: ```suggestion
**Example backtrace**
```

    ],
    package='slam_toolbox',
    executable='sync_slam_toolbox_node',
    name='slam_toolbox',
    prefix=['xterm -e gdb -ex run --args'],  # For interactive GDB in a separate window/GUI
    output='screen')

- **Remote Debugging (without GUI):** If debugging without GUI, omit ``xterm -e`` :

.. code-block:: bash

  prefix=['gdb -ex run --args']

GDB's output and interaction will happen within the terminal session where you launched the ROS 2 application.
Here's an similar example for the ``'start_sync_slam_toolbox_node'`` -

.. code-block:: python

  start_sync_slam_toolbox_node = Node(
    parameters=[
        get_package_share_directory("slam_toolbox") + '/config/mapper_params_online_sync.yaml',
        {'use_sim_time': use_sim_time}
    ],
    package='slam_toolbox',
    executable='sync_slam_toolbox_node',
    name='slam_toolbox',
    prefix=['gdb -ex run --args'],  # For GDB within the launch terminal
    output='screen')

Just as before, this prefix will launch a GDB session, now in ``xterm`` and run the launch file you requested with all the additional launch arguments defined.

Once your node crashes, you'll see a prompt like below, now in the ``xterm`` session. 

> At this point you can now get a backtrace, Further follow `Reading the Stack Trace`_.

Reviewer: ```suggestion
At this point you can now get a backtrace, and read it using the instructions in `Reading the Stack Trace`.
```

If your node of interest is being launched from a nested launch file (e.g. an included launch file) you may want to do the following:

- Comment out the launch file inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

- Launch the node's launch file in another terminal following the instructions in `From a Launch File`_.

Alternatively, if your node of interest is being launched in these files directly (e.g. you see a ``Node``, ``LifecycleNode``, or inside a ``ComponentContainer``), you will need to seperate this from the others:

- Comment out the node's inclusion from the parent launch file

- Recompile the package of interest with ``-g`` flag for debug symbols

- Launch the parent launch file in a terminal

- Launch the server's node in another terminal following the instructions in `Debugging a specific node with GDB`_.

.. note::

  In this case you may need to remap or provide parameter files to this node if it was previously provided by the launch file.
  Using ``--ros-args`` you can give it the path to the new parameters file, remaps, or names.

>   See :doc:`this tutorial <../../Guides/Node-arguments.html>` for the commandline arguments required.

Reviewer: ```suggestion
  See :doc:`this tutorial <../../How-To-Guides/Node-arguments>` for the commandline arguments required.
```

  We understand this can be a pain, so it might encourage you to rather have each node possible as a separately included launch file to make debugging easier. 
  An example set of arguments might be ``--ros-args -r __node:=<node_name> --params-file /absolute/path/to/params.yaml`` (as a template).

Once your node crashes, you'll see a prompt like below in the specific server's terminal. 

> At this point you can now get a backtrace, Further follow `Reading the Stack Trace`_.

Reviewer: ```suggestion
At this point you can now get a backtrace, and read it using the instructions in `Reading the Stack Trace`.
```

Debugging tests with GDB
========================

If a C++ test is failing, GDB can be used directly on the test executable in the build directory.
Ensure to build the code in debug mode.
Since the previous build type may be cached by CMake, clean the cache and rebuild.

.. code-block:: console

  colcon build --cmake-clean-cache --mixin debug

In order for GDB to load debug symbols for any shared libraries called, make sure to source your environment.
This configures the value of ``LD_LIBRARY_PATH``.

.. code-block:: console

  source install/setup.bash

Finally, run the test directly through GDB.
For example:

.. code-block:: console

  gdb -ex run ./build/rcl/test/test_logging

If the code is throwing an unhandled exception, you can catch it in GDB before gtest handles it.

.. code-block:: console

  gdb ./build/rcl/test/test_logging
  catch throw
  run

Automatic backtrace on crash
============================

The `backward-cpp <https://github.com/pal-robotics/backward_ros>`_ library provides beautiful stack traces, and the `backward_ros <https://github.com/pal-robotics/backward_ros>`_ wrapper simplifies its integration.

> Just add it as a dependency and `find_package` it in your CMakeLists and the backward libraries will be injected in all your executables and libraries.

Reviewer: ```suggestion
Just add it as a dependency and ``find_package`` it in your CMakeLists and the backward libraries will be injected in all your executables and libraries.
```
